Library/Paginator.swift

```diff
@@ -54,84 +118,75 @@ public class Paginator<Envelope, Value: Equatable, Cursor, SomeError: Error, Req
     requestFromParams: @escaping ((RequestParams) -> AnyPublisher<Envelope, SomeError>),
     requestFromCursor: @escaping ((Cursor) -> AnyPublisher<Envelope, SomeError>)
   ) {
-    self.values = []
-    self.isLoading = false
-    self.error = nil
-    self.state = .unloaded
+    self.results = .unloaded
 
     self.valuesFromEnvelope = valuesFromEnvelope
     self.cursorFromEnvelope = cursorFromEnvelope
     self.requestFromParams = requestFromParams
     self.requestFromCursor = requestFromCursor
   }
 
-  func handleRequest(_ request: AnyPublisher<Envelope, SomeError>) {
+  func handleRequest(_ request: AnyPublisher<Envelope, SomeError>, shouldClear: Bool) {
     request
-      .receive(on: RunLoop.main)
-      .catch { [weak self] error -> AnyPublisher<Envelope, SomeError> in
-        self?.error = error
-        self?.state = .error
-        return Empty<Envelope, SomeError>().eraseToAnyPublisher()
-      }
-      .assertNoFailure()
-      .handleEvents(receiveCompletion: { [weak self] _ in
-        self?.isLoading = false
-      }, receiveCancel: { [weak self] in
-        self?.isLoading = false
-      })
-      .sink(receiveValue: { [weak self] envelope in
-        guard let self else { return }
+      .combineLatest(self.$results.first().setFailureType(to: SomeError.self))
+      .map { [weak self] envelope, previousResults -> Results in
+        guard let self else {
+          fatalError()
+        }
 
         let newValues = self.valuesFromEnvelope(envelope)
-        self.values.append(contentsOf: newValues)
-
-        let cursor = self.cursorFromEnvelope(envelope)
-        self.lastCursor = cursor
-
-        if self.values.count == 0 {
-          self.state = .empty
-        } else if cursor == nil || newValues.count == 0 {
-          self.state = .allLoaded
+        let nextCursor = self.cursorFromEnvelope(envelope)
+        var allValues = shouldClear ? [] : previousResults.values
+        allValues.append(contentsOf: newValues)
+
+        let results: Results = if allValues.count == 0 {
```

Comment:
nit: I haven't seen this this kind of syntax before and I find it kind of hard to read. Would you be opposed to rewriting this as a typical nested `else if` thing with early returns instead?

Reply:
Sure; this is [new syntax](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0380-if-switch-expressions.md) as of Swift 5.9.